File: vscode-extension/src/contracts/snapshot.ts
export type Period = 'today' | '7d' | '30d' | 'month';

export interface SnapshotFilters {
  from?: string | null;
  to?: string | null;
  period?: Period | null;
  project?: string | null;
  model?: string | null;
  source?: string | null;
  user?: string | null;
  machine?: string | null;
}

export interface SnapshotSummary {
  sessions: number;
  total_tokens: number;
  tokens_saved: number;
  cache_ratio: number | null;
  observed_cost_usd: number | null;
  estimated_cost_usd: number | null;
  estimated_savings_usd: number | null;
}

export type BillingMode = 'api' | 'chatgpt_codex_plan' | 'unknown' | 'mixed';
export type BillingConfidence = 'explicit' | 'unknown' | 'mixed';
export type EstimatedCostBasis = 'openai_api_estimate' | 'openai_api_equivalent';

export interface SnapshotBilling {
  mode: BillingMode;
  confidence: BillingConfidence;
  estimated_cost_basis: EstimatedCostBasis;
  is_observed_spend: boolean;
}

export interface SnapshotFreshness {
  last_imported_at: string | null;
  artifacts_tracked: number;
}

export type AvailabilityReason =
  | 'source_does_not_report_cost'
  | 'insufficient_pricing_data'
  | 'no_optimization_data';

export interface AvailabilityItem {
  available: boolean;
  reason: AvailabilityReason | null;
}

export interface SnapshotAvailability {
  observed_cost: AvailabilityItem;
  estimated_cost: AvailabilityItem;
  estimated_savings: AvailabilityItem;
}

export interface DailySeriesPoint {
  date: string;
  sessions: number;
  total_tokens: number;
  cache_ratio: number | null;
  observed_cost_usd: number | null;
  estimated_cost_usd: number | null;
  estimated_savings_usd: number | null;
}

export interface ProjectBreakdown {
  project: string;
  sessions: number;
  total_tokens: number;
}

export interface ModelBreakdown {
  model: string;
  sessions: number;
  total_tokens: number;
  estimated_cost_usd: number | null;
  cost_events_total: number;
  cost_events_priced: number;
  cost_complete: boolean;
}

export interface SourceBreakdown {
  source: string;
  sessions: number;
  total_tokens: number;
}

export interface ClientBreakdown {
  client: string;
  sessions: number;
  total_tokens: number;
  share: number;
}

export interface SnapshotSeries {
  daily: DailySeriesPoint[];
}

export interface SnapshotBreakdowns {
  projects: ProjectBreakdown[];
  models: ModelBreakdown[];
  sources: SourceBreakdown[];
  clients?: ClientBreakdown[];
}

export interface SnapshotDimensions {
  projects: string[];
  models: string[];
  sources: string[];
  users: string[];
  machines: string[];
}

export interface SnapshotQuality {
  import_errors: number;
  tokens_without_model: number;
  identity_confidence: Record<string, number>;
  correlation_confidence: Record<string, number>;
}

export interface ExtensionSnapshot {
  schema: 'agentscope-extension-snapshot';
  version: 2;
  generated_at: string;
  database: string;
  freshness?: SnapshotFreshness;
  filters: SnapshotFilters;
  summary: SnapshotSummary;
  billing: SnapshotBilling;
  availability: SnapshotAvailability;
  series: SnapshotSeries;
  breakdowns: SnapshotBreakdowns;
  dimensions: SnapshotDimensions;
  quality: SnapshotQuality;
}

export type SnapshotContractErrorCode =
  | 'SNAPSHOT_INVALID_JSON'
  | 'SNAPSHOT_UNSUPPORTED_VERSION';

export class SnapshotContractError extends Error {
  constructor(
    public readonly code: SnapshotContractErrorCode,
    message: string,
  ) {
    super(message);
    this.name = 'SnapshotContractError';
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value);
}

function isNumberOrNull(value: unknown): value is number | null {
  return value === null || isFiniteNumber(value);
}

function isStringArray(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === 'string');
}

function isNumberRecord(value: unknown): value is Record<string, number> {
  return isRecord(value) && Object.values(value).every(isFiniteNumber);
}

function invalid(message: string): never {
  throw new SnapshotContractError('SNAPSHOT_INVALID_JSON', message);
}

function validateAvailabilityItem(value: unknown): void {
  const reasons = new Set([
    'source_does_not_report_cost',
    'insufficient_pricing_data',
    'no_optimization_data',
  ]);
  if (!isRecord(value) || typeof value.available !== 'boolean') {
    invalid('Snapshot availability item is invalid.');
  }
  if (value.reason !== null && !reasons.has(String(value.reason))) {
    invalid('Snapshot availability reason is invalid.');
  }
}

function validateBilling(value: unknown): void {
  if (!isRecord(value)) invalid('Snapshot billing section is invalid.');
  const modes = new Set(['api', 'chatgpt_codex_plan', 'unknown', 'mixed']);
  const confidences = new Set(['explicit', 'unknown', 'mixed']);
  const bases = new Set(['openai_api_estimate', 'openai_api_equivalent']);
  if (
    !modes.has(String(value.mode)) ||
    !confidences.has(String(value.confidence)) ||
    !bases.has(String(value.estimated_cost_basis)) ||
    typeof value.is_observed_spend !== 'boolean'
  ) {
    invalid('Snapshot billing section is invalid.');
  }
}

function validateDaily(value: unknown): void {
  if (!Array.isArray(value)) invalid('Snapshot daily series is invalid.');
  for (const item of value) {
    if (
      !isRecord(item) ||
      typeof item.date !== 'string' ||
      !isFiniteNumber(item.sessions) ||
      !isFiniteNumber(item.total_tokens) ||
      !isNumberOrNull(item.cache_ratio) ||
      !isNumberOrNull(item.observed_cost_usd) ||
      !isNumberOrNull(item.estimated_cost_usd) ||
      !isNumberOrNull(item.estimated_savings_usd)
    ) {
      invalid('Snapshot daily series item is invalid.');
    }
  }
}

function validateBreakdown(value: unknown, labelKey: string): void {
  if (!Array.isArray(value)) invalid('Snapshot breakdown is invalid.');
  for (const item of value) {
    if (
      !isRecord(item) ||
      typeof item[labelKey] !== 'string' ||
      !isFiniteNumber(item.sessions) ||
      !isFiniteNumber(item.total_tokens)
    ) {
      invalid(`Snapshot ${labelKey} breakdown is invalid.`);
    }
  }
}

function validateModelBreakdown(value: unknown): void {
  if (!Array.isArray(value)) invalid('Snapshot model breakdown is invalid.');
  for (const item of value) {
    if (
      !isRecord(item) ||
      typeof item.model !== 'string' ||
      !isFiniteNumber(item.sessions) ||
      !isFiniteNumber(item.total_tokens) ||
      !isNumberOrNull(item.estimated_cost_usd) ||
      !isFiniteNumber(item.cost_events_total) ||
      !isFiniteNumber(item.cost_events_priced) ||
      typeof item.cost_complete !== 'boolean'
    ) {
      invalid('Snapshot model breakdown is invalid.');
    }
    if (
      item.cost_events_total < 0 ||
      item.cost_events_priced < 0 ||
      item.cost_events_priced > item.cost_events_total
    ) {
      invalid('Snapshot model cost coverage is invalid.');
    }
  }
}

function validateClientBreakdown(value: unknown): void {
  if (!Array.isArray(value)) invalid('Snapshot client breakdown is invalid.');
  for (const item of value) {
    if (
      !isRecord(item) ||
      typeof item.client !== 'string' ||
      !isFiniteNumber(item.sessions) ||
      !isFiniteNumber(item.total_tokens) ||
      !isFiniteNumber(item.share) ||
      item.share < 0 ||
      item.share > 1
    ) {
      invalid('Snapshot client breakdown is invalid.');
    }
  }
}

function validateFreshness(value: unknown): void {
  if (!isRecord(value)) invalid('Snapshot freshness is invalid.');
  if (
    value.last_imported_at !== null &&
    typeof value.last_imported_at !== 'string'
  ) {
    invalid('Snapshot last import timestamp is invalid.');
  }
  if (!isFiniteNumber(value.artifacts_tracked)) {
    invalid('Snapshot tracked artifact count is invalid.');
  }
}

export function parseExtensionSnapshot(value: unknown): ExtensionSnapshot {
  if (!isRecord(value)) return invalid('Snapshot must be an object.');
  if (value.schema !== 'agentscope-extension-snapshot') return invalid('Unexpected snapshot schema.');
  if (value.version !== 2) {
    throw new SnapshotContractError(
      'SNAPSHOT_UNSUPPORTED_VERSION',
      `Unsupported snapshot version: ${String(value.version)}`,
    );
  }
  if (typeof value.generated_at !== 'string' || typeof value.database !== 'string') {
    return invalid('Snapshot metadata is invalid.');
  }
  if (value.freshness !== undefined) validateFreshness(value.freshness);
  if (
    !isRecord(value.filters) || !isRecord(value.summary) ||
    !isRecord(value.billing) || !isRecord(value.availability) || !isRecord(value.series) ||
    !isRecord(value.breakdowns) || !isRecord(value.dimensions) ||
    !isRecord(value.quality)
  ) {
    return invalid('Snapshot sections are invalid.');
  }

  const summary = value.summary;
  if (
    !isFiniteNumber(summary.sessions) ||
    !isFiniteNumber(summary.total_tokens) ||
    !isFiniteNumber(summary.tokens_saved) ||
    !isNumberOrNull(summary.cache_ratio) ||
    !isNumberOrNull(summary.observed_cost_usd) ||
    !isNumberOrNull(summary.estimated_cost_usd) ||
    !isNumberOrNull(summary.estimated_savings_usd)
  ) {
    return invalid('Snapshot summary is invalid.');
  }

  validateBilling(value.billing);
  validateAvailabilityItem(value.availability.observed_cost);
  validateAvailabilityItem(value.availability.estimated_cost);
  validateAvailabilityItem(value.availability.estimated_savings);
  validateDaily(value.series.daily);
  validateBreakdown(value.breakdowns.projects, 'project');
  validateModelBreakdown(value.breakdowns.models);
  validateBreakdown(value.breakdowns.sources, 'source');
  if (value.breakdowns.clients !== undefined) {
    validateClientBreakdown(value.breakdowns.clients);
  }

  const dimensions = value.dimensions;
  if (
    !isStringArray(dimensions.projects) ||
    !isStringArray(dimensions.models) ||
    !isStringArray(dimensions.sources) ||
    !isStringArray(dimensions.users) ||
    !isStringArray(dimensions.machines)
  ) {
    return invalid('Snapshot dimensions are invalid.');
  }

  const quality = value.quality;
  if (
    !isFiniteNumber(quality.import_errors) ||
    !isFiniteNumber(quality.tokens_without_model) ||
    !isNumberRecord(quality.identity_confidence) ||
    !isNumberRecord(quality.correlation_confidence)
  ) {
    return invalid('Snapshot quality section is invalid.');
  }

  const allowedPeriods = new Set(['today', '7d', '30d', 'month']);
  for (const key of ['from', 'to', 'project', 'model', 'source', 'user', 'machine'] as const) {
    const filterValue = value.filters[key];
    if (filterValue !== undefined && filterValue !== null && typeof filterValue !== 'string') {
      return invalid(`Invalid filter: ${key}`);
    }
  }
  const period = value.filters.period;
  if (period !== undefined && period !== null && !allowedPeriods.has(String(period))) {
    return invalid('Invalid period filter.');
  }

  return value as unknown as ExtensionSnapshot;
}

File: src/agentscope/extension/snapshot.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path

from agentscope.analytics.dashboard import DashboardAnalyticsService
from agentscope.analytics.filters import AnalyticsFilter
from agentscope.billing import billing_semantics
from agentscope.extension.contracts import (
    SNAPSHOT_SCHEMA,
    SNAPSHOT_VERSION,
    AvailabilityItem,
    SnapshotAvailability,
    SnapshotBilling,
    SnapshotDimensions,
    SnapshotQuality,
    SnapshotSummary,
    to_dict,
)
from agentscope.storage.repository import Repository
from agentscope.usage_context import ensure_usage_context_schema


def _dimension_values(repository: Repository, table: str, expression: str) -> list[str]:
    with repository.database.connect() as conn:
        rows = conn.execute(
            f"SELECT DISTINCT {expression} AS value FROM {table} "
            "WHERE value IS NOT NULL AND trim(value) <> '' ORDER BY value"
        ).fetchall()
    return [str(row["value"]) for row in rows]


def _dimensions(repository: Repository) -> SnapshotDimensions:
    return SnapshotDimensions(
        projects=_dimension_values(repository, "projects", "name"),
        models=_dimension_values(repository, "models", "name"),
        sources=_dimension_values(repository, "sources", "name"),
        users=_dimension_values(
            repository,
            "users",
            "COALESCE(display_name, stable_key)",
        ),
        machines=_dimension_values(
            repository,
            "machines",
            "COALESCE(display_name, stable_key)",
        ),
    )


def _freshness(repository: Repository) -> dict[str, object]:
    with repository.database.connect() as conn:
        row = conn.execute(
            """
            SELECT MAX(last_imported_at) AS last_imported_at,
                   COUNT(*) AS artifacts_tracked
            FROM import_state
            WHERE status='complete'
            """
        ).fetchone()
    return {
        "last_imported_at": (
            str(row["last_imported_at"])
            if row["last_imported_at"] is not None
            else None
        ),
        "artifacts_tracked": int(row["artifacts_tracked"] or 0),
    }


def _identity_confidence(repository: Repository) -> dict[str, int]:
    with repository.database.connect() as conn:
        rows = conn.execute(
            """
            SELECT identity_confidence, COUNT(*) AS n
            FROM users
            GROUP BY identity_confidence
            ORDER BY identity_confidence
            """
        ).fetchall()
    return {str(row["identity_confidence"]): int(row["n"]) for row in rows}


def _scope_clauses(
    filters: AnalyticsFilter,
    *,
    date_expression: str,
    model_expression: str,
) -> tuple[list[str], list[object]]:
    clauses: list[str] = []
    params: list[object] = []
    local_day = filters.local_date_expression(date_expression)
    if filters.from_date is not None:
        clauses.append(f"{local_day} >= ?")
        params.append(filters.from_date.isoformat())
    if filters.to_date is not None:
        clauses.append(f"{local_day} <= ?")
        params.append(filters.to_date.isoformat())
    if filters.project is not None:
        clauses.append("p.name = ?")
        params.append(filters.project)
    if filters.source is not None:
        clauses.append("src.name = ?")
        params.append(filters.source)
    if filters.user is not None:
        clauses.append("COALESCE(u.display_name, u.stable_key) = ?")
        params.append(filters.user)
    if filters.machine is not None:
        clauses.append("COALESCE(mc.display_name, mc.stable_key) = ?")
        params.append(filters.machine)
    if filters.model is not None:
        clauses.append(f"{model_expression} = ?")
        params.append(filters.model)
    return clauses, params


def _tokens_without_model(repository: Repository, filters: AnalyticsFilter) -> int:
    clauses, params = _scope_clauses(
        filters,
        date_expression="tu.timestamp",
        model_expression="COALESCE(tm.name, sm.name)",
    )
    clauses.insert(0, "COALESCE(tu.model_id, s.model_id) IS NULL")
    with repository.database.connect() as conn:
        row = conn.execute(
            """
            SELECT COUNT(*) AS n
            FROM token_usage tu
            JOIN sessions s ON s.id=tu.session_id
            LEFT JOIN projects p ON p.id=s.project_id
            LEFT JOIN models tm ON tm.id=tu.model_id
            LEFT JOIN models sm ON sm.id=s.model_id
            JOIN sources src ON src.id=s.source_id
            LEFT JOIN users u ON u.id=s.user_id
            LEFT JOIN machines mc ON mc.id=s.machine_id
            WHERE """ + " AND ".join(clauses),
            params,
        ).fetchone()
    return int(row["n"])


def _has_savings_evidence(repository: Repository, filters: AnalyticsFilter) -> bool:
    cost_clauses, cost_params = _scope_clauses(
        filters,
        date_expression="c.period_start",
        model_expression="COALESCE(cm.name, sm.name)",
    )
    cost_clauses.insert(
        0,
        "(c.compression_savings_usd IS NOT NULL "
        "OR c.cache_savings_usd IS NOT NULL "
        "OR c.total_savings_usd IS NOT NULL)",
    )
    optimization_clauses, optimization_params = _scope_clauses(
        filters,
        date_expression="op.timestamp",
        model_expression="COALESCE(om.name, sm.name)",
    )
    optimization_clauses.insert(
        0,
        "(op.compression_savings_usd IS NOT NULL OR op.cache_savings_usd IS NOT NULL)",
    )
    with repository.database.connect() as conn:
        cost_row = conn.execute(
            """
            SELECT COUNT(*) AS n
            FROM costs c
            LEFT JOIN sessions s ON s.id=c.session_id
            LEFT JOIN projects p ON p.id=s.project_id
            LEFT JOIN models cm ON cm.id=c.model_id
            LEFT JOIN models sm ON sm.id=s.model_id
            LEFT JOIN sources src ON src.id=s.source_id
            LEFT JOIN users u ON u.id=s.user_id
            LEFT JOIN machines mc ON mc.id=s.machine_id
            WHERE """ + " AND ".join(cost_clauses),
            cost_params,
        ).fetchone()
        if int(cost_row["n"]) > 0:
            return True
        optimization_row = conn.execute(
            """
            SELECT COUNT(*) AS n
            FROM optimizations op
            LEFT JOIN sessions s ON s.id=op.session_id
            LEFT JOIN projects p ON p.id=s.project_id
            LEFT JOIN models om ON om.id=op.model_id
            LEFT JOIN models sm ON sm.id=s.model_id
            LEFT JOIN sources src ON src.id=s.source_id
            LEFT JOIN users u ON u.id=s.user_id
            LEFT JOIN machines mc ON mc.id=s.machine_id
            WHERE """ + " AND ".join(optimization_clauses),
            optimization_params,
        ).fetchone()
    return int(optimization_row["n"]) > 0


def _availability(value: float | None, reason: str) -> AvailabilityItem:
    return AvailabilityItem(
        available=value is not None,
        reason=None if value is not None else reason,
    )


def build_extension_snapshot(
    repository: Repository,
    filters: AnalyticsFilter,
    *,
    period: str | None,
    database_path: Path,
) -> dict[str, object]:
    ensure_usage_context_schema(repository)
    dashboard = DashboardAnalyticsService(repository, filters)
    summary = dashboard.summary()
    quality = dashboard.data_quality()
    optimization_confidence = quality.get("optimization_confidence", {})
    if not isinstance(optimization_confidence, dict):
        optimization_confidence = {}

    observed_cost = summary.observed_cost_usd
    estimated_cost = summary.estimated_raw_cost_usd
    estimated_savings = (
        summary.total_savings_usd
        if _has_savings_evidence(repository, filters)
        else None
    )
    availability = SnapshotAvailability(
        observed_cost=_availability(observed_cost, "source_does_not_report_cost"),
        estimated_cost=_availability(estimated_cost, "insufficient_pricing_data"),
        estimated_savings=_availability(estimated_savings, "no_optimization_data"),
    )
    billing = billing_semantics(repository, filters)

    return {
        "schema": SNAPSHOT_SCHEMA,
        "version": SNAPSHOT_VERSION,
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "database": str(database_path),
        "freshness": _freshness(repository),
        "filters": {
            "from": filters.from_date.isoformat() if filters.from_date else None,
            "to": filters.to_date.isoformat() if filters.to_date else None,
            "period": period,
            "project": filters.project,
            "model": filters.model,
            "source": filters.source,
            "user": filters.user,
            "machine": filters.machine,
        },
        "summary": to_dict(
            SnapshotSummary(
                sessions=summary.sessions,
                total_tokens=summary.total_tokens,
                tokens_saved=summary.tokens_saved,
                cache_ratio=summary.cache_ratio if summary.input_tokens else None,
                observed_cost_usd=observed_cost,
                estimated_cost_usd=estimated_cost,
                estimated_savings_usd=estimated_savings,
            )
        ),
        "billing": to_dict(
            SnapshotBilling(
                mode=billing.mode,
                confidence=billing.confidence,
                estimated_cost_basis=billing.estimated_cost_basis,
                is_observed_spend=billing.is_observed_spend,
            )
        ),
        "availability": to_dict(availability),
        "series": {"daily": dashboard.by_day()},
        "breakdowns": {
            "projects": dashboard.by_project(),
            "models": dashboard.by_model(),
            "sources": dashboard.by_source(),
            "clients": dashboard.by_client(),
        },
        "dimensions": to_dict(_dimensions(repository)),
        "quality": to_dict(
            SnapshotQuality(
                import_errors=int(quality.get("import_errors", 0) or 0),
                tokens_without_model=_tokens_without_model(repository, filters),
                identity_confidence=_identity_confidence(repository),
                correlation_confidence={
                    str(key): int(value)
                    for key, value in optimization_confidence.items()
                },
            )
        ),
    }
